service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AllergyStructureMapper.java

```diff
@@ -188,6 +190,16 @@ private String buildNotePertinentInformation(AllergyIntolerance allergyIntoleran
                 .map(Annotation::getText)
                 .collect(Collectors.joining(StringUtils.SPACE));
         }
+        List<Condition> relatedConditions = messageContext.getInputBundleHolder().getRelatedConditions(allergyIntolerance.getId());
+        for (var relatedCondition: relatedConditions) {
+            for (var annotation: relatedCondition.getNote()) {
+                if (notes.equals(StringUtils.EMPTY)) {
+                    notes = annotation.getText();
+                } else {
+                    notes = StringUtils.joinWith(StringUtils.SPACE, notes, annotation.getText());
+                }
+            }
+        }
```

Comment:
First observation is that the `if` `else` branches are unnecessary

```suggestion
        List<Condition> relatedConditions = messageContext.getInputBundleHolder().getRelatedConditions(allergyIntolerance.getId());
        for (var relatedCondition: relatedConditions) {
            for (var annotation: relatedCondition.getNote()) {
                notes = StringUtils.joinWith(StringUtils.SPACE, notes, annotation.getText());
            }
        }
```

Now it's clear this could be a streaming operation, working on the collection as a whole rather than manually iterating.

```suggestion
        String relatedConditions = messageContext.getInputBundleHolder().getRelatedConditions(allergyIntolerance.getId())
            .stream()
            .map(RelatedConditionClass::getNote)
            .flatMap(List::stream)
            .collect(Collectors.joining(" "));
       notes = StringUtils.joinWith(StringUtils.SPACE, notes, relatedConditions);
```

But the stream is being treated the same in both cases, _and_ the results of the two streams are combined in the same way, so we could get a single stream of the relevant `Annotation`s from both sources with `Stream.concat()` and do the work all at once:

(whole method body)
```suggestion
    return Stream.concat(
        // stream of Annotations from the first part of the method
        allergyIntolerance.hasNote() ? allergyIntolerance.getNote().stream() : Stream.empty(),
        
        // stream of Annotations from the new part
        messageContext.getInputBundleHolder().getRelatedConditions(allergyIntolerance.getId())
            .stream()
            .map(RelatedConditionClass::getNote)
            .flatMap(List::stream)
    )
        // combine them all together
        .map(Annotation::getText)
        .collect(Collectors.joining(StringUtils.SPACE));
```

Reply:
Also applies to [`ImmunizationObservationStatementMapper#buildNotePertinentInformation`](https://github.com/nhsconnect/integration-adaptor-gp2gp/pull/140/files#diff-b4a7d5f85c231e5758534b0c623f237217436ddc03228ea31e8a73bd281f759aR190-R199)

Reply:
Perfect! Thank you, @andrewwa-kainos! It definitely looks much better!
3b29cbb718a07d951b35faff10b5d131e17e9350